atoma-proxy/src/server/middleware.rs

```diff
@@ -1962,17 +1994,54 @@ pub mod auth {
                 .await;
         };
 
-        // NOTE: At this point, we have an acquired stack lock, so we can safely acquire a new stack.
-        acquire_new_stack(
-            state.state_manager_sender.clone(),
-            user_id,
-            lock_guard,
-            endpoint.to_string(),
-            total_tokens,
-            Arc::clone(&state.sui),
-            node,
-        )
-        .await
+        // We don't have a stack for the user, lets check if the user is using fiat currency.
+        let (result_sender, result_receiver) = oneshot::channel();
+        state
+            .state_manager_sender
+            .send(AtomaAtomaStateManagerEvent::LockUserFiatBalance {
+                user_id,
+                amount: total_tokens as i64 * node.price_per_one_million_compute_units as i64
+                    / ONE_MILLION as i64,
+                result_sender,
+            })
+            .map_err(|err| AtomaProxyError::InternalError {
+                message: format!("Failed to send LockUserFiatBalance event: {err:?}"),
+                client_message: None,
+                endpoint: endpoint.to_string(),
+            })?;
+
+        let locked_fiat = result_receiver
+            .await
+            .map_err(|err| AtomaProxyError::InternalError {
+                message: format!("Failed to receive LockUserFiatBalance result: {err:?}"),
+                client_message: None,
+                endpoint: endpoint.to_string(),
+            })?
+            .map_err(|err| AtomaProxyError::InternalError {
+                message: format!("Failed to get LockUserFiatBalance result: {err:?}"),
+                client_message: None,
+                endpoint: endpoint.to_string(),
+            })?;
+
+        if locked_fiat {
```

Comment:
How is the user's fiat balance unlocked? wouldn't this be after they have used up all their tokens?